opm/simulators/wells/WellInterfaceFluidSystem.cpp

```diff
@@ -236,10 +256,28 @@ activeInjectionConstraint(const SingleWellState& ws,
 
     if (controls.hasControl(Well::InjectorCMode::THP) && currentControl != Well::InjectorCMode::THP)
     {
-        const auto& thp = getTHPConstraint(summaryState);
+        const auto& thp_limit = getTHPConstraint(summaryState);
         double current_thp = ws.thp;
-        if (thp < current_thp)
-            return Well::InjectorCMode::THP;
+        if (thp_limit < current_thp) {
+            bool increase = true;
+            for (int p = 0; p < number_of_phases_; ++p) {
+                // Currently we use the well potentials here computed before the iterations.
+                // We may need to recompute the well potentials to get a more
+                // accurate check here.
+                increase = increase && (ws.surface_rates[p]) >= ws.well_potentials[p];
+            }
+            if(!increase) {
```

Comment:
`increase==true` means well injects less under THP control, we should switch to THP control. Can you double check this logic here. 

The reason we do not want to switch to THP control is because the current control is the more restrictive one than THP control, means the current inject rates is smaller than the rates under THP target (`increase == false`).

Reply:
And also, for injectors, should we compare the injection phase in case there are some noise for the other near-zero values?

Reply:
> increase==true means well injects less under THP control, we should switch to THP control. Can you double check this logic here.

The logic for an injector is the opposite of what you have for an producer. That is reflected in the code as it is. 

> And also, for injectors, should we compare the injection phase in case there are some noise for the other near-zero values?

Your point is that even if the rate of the injection phase is increasing there may be oscillation(production) in the other phases (i.e. ws.surface_rates[p] < 0) that could allow for switching to thp. That certainly is possible, but I have not seen it in the cases I have looked at. I can change the code accordingly, by adding switch(injectorType).... case.... around the check. Or we can cross that bridge when we come there. It is hard to guess what is the robust logic here without any cases where the feature is exposed.

Reply:
> The logic for an injector is the opposite of what you have for an producer. That is reflected in the code as it is.

But the well control logic is that we stay at the more restrictive control mode (produce or inject less). If you agree with this logic, then the code for producers and injectors should be the same. And the code for the injectors is basically saying the opposite to the logic.

Reply:
I think I finally got your point. I have update the PR accordingly. Still doing some testing, to see the effect of the change for the injector. My main test case only expose the production part.

Reply:
I was very confused now. I think the "increase" variable name is very unfortunate.

I think the code itself is correct now: If the current rate is less than the potential, then `increase` will be true (this is very weird... call it `rate_less_than_potential` to be clearer). If `increase` is true we will enter the `else` branch and not switch.

Reply:
the logic with the naming from my side was that if rate < pot the rate will increase if we switch to thp. i.e. only switch if !increase. `rate_less_than_potential`  is a more descriptive name. I will update accordingly.